README
NAME
    Switcheroo - yet another switch statement for Perl

SYNOPSIS
       my $day_type;
   
       switch ($day) {
          case 0, 6:  $day_type = "weekend";
          default:    $day_type = "weekday";
       }

STATUS
    Experimental.

    No backwards compatibility between releases is guaranteed. Both the
    surface syntax and the internals of the module are liable to change at my
    whim.

DESCRIPTION
    This module provides Perl with a switch statement. It's more reliable than
    the Switch module (which is broken on recent versions of Perl anyway),
    less confusing than `use feature 'switch'`, and more powerful than
    Switch::Plain (though Switch::Plain is significantly faster).

    Switcheroo uses the Perl keyword API, which was introduced in Perl 5.14,
    so this module does not work on older releases of Perl.

    The basic grammar of the switch statement is as follows:

       switch ( TEST ) {
          case EXPR1: STATEMENT1;
          case EXPR2: STATEMENT2;
          default:    STATEMENT3;
       }

    TEST is evaluated in scalar context. Each expression EXPR1, EXPR2, etc is
    evaluated in list context. If TEST matches any of the expression, then the
    statement following it is executed. Matching is performed by
    match::simple, which is a simplified version of the Perl smart match
    operator. If no match is successful, then the `default` statement is
    executed.

    `switch` is whole statement, so does not need to be followed by a
    semicolon.

    Within the switch block, $_ is a read-only alias to the TEST value.

    That's the basics taken care of, but there are several variations...

  Implicit test
    If the test is omitted, then $_ is tested:

       my $day_type;
   
       $_ = $day;
       switch {
          case 0, 6:  $day_type = "weekend";
          default:    $day_type = "weekday";
       }

  Expression blocks
    If `case` is followed by a `{` character, this is *not* interpreted as the
    start of an anonymous hashref, but as a block. Matching via match::simple
    is not attempted; instead the block is evaluated as a boolean.

       switch ($number) {
          case 0:           say "zero";
          case { $_ % 2 }:  say "an odd number";
          default:          say "an even number";
       }

  Regexp Expressions
    If this module encounters:

       switch ($foo) { case /foo/: say "foo" }

    Don't worry; we know you meant `qr/foo/` and not `m/foo/`.

  Statement blocks
    If the first non-whitespace character is `{`, the statement is treated as
    a block rather than a single statement:

       switch ($number) {
          case 0: {
             say "zero";
          }
          case { $_ % 2 }: {
             say "an odd number";
          }
          default: {
             say "an even number";
          }
       }

  Comparison expression
    Above I said that matching is performed by match::simple. That was a lie.
    match::simple is just the default. You can provide your own expression for
    matching:

       switch ($number) mode ($a > $b) {
          case 1000:   say "greater than 1000";
          case 100:    say "greater than 100";
          case 10:     say "greater than 10";
          case 1:      say "greater than 1";
       }

    $a is the TERM and $b is the EXPR. These are the same special package
    variables used by `sort` and by `reduce` from List::Util.

  Switch expressions
    Although `switch` acts as a full statement usually, it can be used as part
    of an expression if the keyword `do` appears before the block:

       my $day_type = switch ($day) do {
          case 0, 6:  "weekend";
          default:    "weekday";
       };

  Fallthrough
    There's no fallthrough.

  `match($x, $y)`
    This module can also re-export the `match` function from match::simple,
    but not by default.

       use Switcheroo qw( match switch );

CAVEATS
    Internally a lot of parts of code are passed around as coderefs, so
    certain things might not work how you'd expect inside `switch`:

    *   `caller`

    *   `return`

    *   @_

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Switcheroo>.

SEE ALSO
    "Switch Statements" in perlsyn, Switch, Switch::Plain.

    <http://en.wikipedia.org/wiki/The_Burning_(Seinfeld)>.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Switcheroo
==========

Created:      2013-09-01
Home page:    <https://metacpan.org/release/Switcheroo>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Switcheroo>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.007	2013-10-10	Bond, James Bond

 [ Test Suite ]
 - More test cases for using @_ within switch blocks.
 - Test cases for closing over variables.

 [ Other ]
 - Improve support for using @_ within switch blocks.

0.006	2013-10-10

 - Export using Exporter::Tiny.
 - Sugar for regexps in case expressions.

0.005	2013-10-10

 [ Packaging ]
 - Rename PerlX::Switch -> Switcheroo.

0.004	2013-09-22

 [ Bug Fixes ]
 - Minor pod syntax fix.

 [ Packaging ]
 - Use a newer version of Dist::Inkt.

0.003	2013-09-22

 [ Documentation ]
 - Document `$a` and `$b` better.
 - Document the status of this distribution.

 [ Other ]
 - Prevent warnings from `$a` and `$b` using a little trick in `import`.
 - The `if` option is now called `mode`.

0.002	2013-09-22

 [ Documentation ]
 - Document Perl 5.14 requirement.
 - Document that the module re-exports `match`.

 [ Test Suite ]
 - Include some test cases for @_ within switch blocks.

 [ Other ]
 - Some improvements accessing @_ within switch blocks, but still... caveat
   invocator!
 - switch now internally uses tail calls (`goto`) when possible.

0.001	2013-09-21	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "yet another switch statement for Perl",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.013, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Switcheroo",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => {
                                       requires => {
                                         "Devel::Caller"   => 0,
                                         "Devel::LexAlias" => 0.05,
                                         "Exporter::Tiny"  => 0.026,
                                         "match::simple"   => 0.004,
                                         "PadWalker"       => 0,
                                         "Parse::Keyword"  => 0.06,
                                         "perl"            => 5.014,
                                       },
                                     },
                        test      => { requires => { "Test::More" => 0.96 } },
                      },
  "provides"       => { Switcheroo => { file => "lib/Switcheroo.pm", version => 0.007 } },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => { web => "http://rt.cpan.org/Dist/Display.html?Queue=Switcheroo" },
                        homepage     => "https://metacpan.org/release/Switcheroo",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => { type => "git", web => "https://github.com/tobyink/p5-switcheroo" },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/Switcheroo/project",
                      },
  "version"        => 0.007,
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

